semtypes/src/main/java/io/ballerina/types/typeops/CellOps.java

```diff
@@ -0,0 +1,209 @@
+    }
+
+    private static boolean cellMutUnlimitedInhabited(Context cx, SemType pos, Conjunction negList) {
+        Conjunction neg = negList;
+        while (neg != null) {
+            if (cellAtomType(neg.atom).mut == CellAtomicType.CellMutability.CELL_MUT_LIMITED &&
+                    Core.isSameType(cx, PredefinedType.TOP, cellAtomType(neg.atom).ty)) {
+                return false;
+            }
+            neg = neg.next;
+        }
+        SemType negListUnionResult = cellNegListUnlimitedUnion(negList);
+        // We expect `isNever` condition to be `true` when there are no negative atoms with unlimited mutability.
+        // Otherwise, we do `isEmpty` to conclude on the inhabitance.
+        return PredefinedType.NEVER.equals(negListUnionResult) || !Core.isEmpty(cx, Core.diff(pos, negListUnionResult));
+    }
+
+    private static SemType cellNegListUnlimitedUnion(Conjunction negList) {
+        SemType negUnion = PredefinedType.NEVER;
+        Conjunction neg = negList;
+        while (neg != null) {
+            if (cellAtomType(neg.atom).mut == CellAtomicType.CellMutability.CELL_MUT_UNLIMITED) {
+                negUnion = Core.union(negUnion, cellAtomType(neg.atom).ty);
+            }
+            neg = neg.next;
+        }
+        return negUnion;
+    }
+
+    private static CellAtomicType intersectCellAtomicType(CellAtomicType c1, CellAtomicType c2) {
+        SemType ty = Core.intersect(c1.ty, c2.ty);
+        CellAtomicType.CellMutability mut = CellAtomicType.CellMutability.fromValue(
+                Integer.min(c1.mut.getValue(), c2.mut.getValue())
+        );
+        return CellAtomicType.from(ty, mut);
+    }
+
+    private static ProperSubtypeData cellSubtypeUnion(SubtypeData t1, SubtypeData t2) {
+        return cellSubtypeDataEnsureProper(bddSubtypeUnion(t1, t2));
+    }
+
+    private static ProperSubtypeData cellSubtypeIntersect(SubtypeData t1, SubtypeData t2) {
+        return cellSubtypeDataEnsureProper(bddSubtypeIntersect(t1, t2));
+    }
+
+    private static ProperSubtypeData cellSubtypeDiff(SubtypeData t1, SubtypeData t2) {
+        return cellSubtypeDataEnsureProper(bddSubtypeDiff(t1, t2));
+    }
+
+    private static ProperSubtypeData cellSubtypeComplement(SubtypeData t) {
+        return cellSubtypeDataEnsureProper(bddSubtypeComplement(t));
+    }
+
+    /**
+     * SubtypeData being a boolean would result in a BasicTypeBitSet which could either be 0 or 1 << BT_CELL.
+     * This is to avoid cell type being a BasicTypeBitSet, as we don't want to lose the cell wrapper.
+     */
+    private static ProperSubtypeData cellSubtypeDataEnsureProper(SubtypeData subtypeData) {
+        if (subtypeData instanceof AllOrNothingSubtype allOrNothingSubtype) {
+            Atom atom;
+            if (allOrNothingSubtype.isAllSubtype()) {
+                atom = ATOM_CELL_VAL;
+            } else {
+                atom = ATOM_CELL_NEVER;
+            }
+            return bddAtom(atom);
```

Comment:
```suggestion
        if (subtypeData instanceof AllOrNothingSubtype allOrNothingSubtype) {
            if (allOrNothingSubtype.isAllSubtype()) {
                return bddAtom(ATOM_CELL_VAL);
            } 
            return bddAtom(ATOM_CELL_NEVER);
```
wdyt?

Reply:
When the logic and number of lines of two branches are the same I usually prefer an `if-else`.  It makes clear that there are two distinct paths of execution which I think is more readable. Having an early return would kind of imply that we are exiting with a specific scenario early which may confuse the reader. 

Additionally, in nballerina we have written the logic the same way here,
https://github.com/ballerina-platform/nballerina/blob/9c70a7cf23d11d2fc568a56ef04a709eefc51456/compiler/modules/types/cell.bal#L153-L159. So again I prefer to have the same style here.  WDYT?

Reply:
+1